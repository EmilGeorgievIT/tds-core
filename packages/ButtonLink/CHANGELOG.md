# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.0.1"></a>
## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-button-link@1.0.0...@tds/core-button-link@1.0.1) (2018-03-22)


### Bug Fixes

* **core-button-link:** standardize focus outline across browsers ([098cb03](https://github.com/telusdigital/tds/commit/098cb03)), closes [#445](https://github.com/telusdigital/tds/issues/445)
