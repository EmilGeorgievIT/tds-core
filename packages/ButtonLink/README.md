# TDS Core: ButtonLink

A link that is styled as a button.
