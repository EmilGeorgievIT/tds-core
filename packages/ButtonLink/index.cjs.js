require('./dist/index.css')
const ButtonLink = require('./dist/index.cjs')

module.exports = ButtonLink
