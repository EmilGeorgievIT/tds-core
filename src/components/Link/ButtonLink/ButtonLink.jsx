import React from 'react'
import PropTypes from 'prop-types'

import { warn } from '../../../warn'
import safeRest from '../../../safeRest'

import styles from './ButtonLink.modules.scss'

/**
 * <span class="docs--badge green">new!</span> <span class="docs--badge purple">v0.21.0</span>
 */
const ButtonLink = ({ reactRouterLinkComponent, variant, children, ...rest }) => {
  if ((reactRouterLinkComponent || rest.to) && !(reactRouterLinkComponent && rest.to)) {
    warn('Link Button', 'The props `reactRouterLinkComponent` and `to` must be used together.')
  }

  return React.createElement(
    reactRouterLinkComponent || 'a',
    {
      ...safeRest(rest),
      className: styles[variant]
    },
    children
  )
}

ButtonLink.propTypes = {
  /**
   * The style.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted']),
  /**
   * React Router Link component.
   */
  reactRouterLinkComponent: PropTypes.func,
  /**
   * Target URL (if using 'reactRouterLinkComponent')
   */
  to: PropTypes.string,
  /**
   * Target URL
   */
  href: PropTypes.string,
  /**
   * The label.
   */
  children: PropTypes.string.isRequired
}
ButtonLink.defaultProps = {
  variant: 'primary',
  reactRouterLinkComponent: null,
  to: null,
  href: null
}

export default ButtonLink
