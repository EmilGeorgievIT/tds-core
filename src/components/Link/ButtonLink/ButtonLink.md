### Minimal Usage

A Button Link is a navigational element that styles itself as a button. Clicking one should navigate to a new "page".

By default, Buttons will be displayed in the `primary` variant. Use primary buttons for the main call to action on a page.

```
<Link.Button href="http://www.telus.com">Find out how</Link.Button>
```

### Using Button Links

All Button Links have the same visual appearance as [Button](#button), but are applied differently in certain contexts. Unlike the Button component, Button Link is more appropriate for:

- Call to action
- Placement atop a complex background such as an image

#### Placing Button Links on images

Use the `outlined` variant when placing a button on top of an image, which will cause the image to show through. You can
also `invert` the colour scheme.

Use this variant with caution. There will be accessibility issues if the colour contrast of the image and the button text
is too low, including the hover state.

```
const Hero = require('../../__docs__/Hero').default;

<Hero>
  <Link.Button variant="outlined" invert>Advanced solutions</Link.Button>
</Hero>
```
