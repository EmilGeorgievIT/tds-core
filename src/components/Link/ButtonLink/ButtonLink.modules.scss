@import '../../../scss/settings/colours';

$button-text-color: $color-white;
$primary-bg-color: $color-primary;
$secondary-bg-color: $color-secondary;

.base {
  text-decoration: none;
}

.primary {
  composes: primary from '../../Button/Button.modules.scss';
  composes: base;

  &:link,
  &:visited {
    color: $button-text-color;
  }
  &:hover {
    color: $primary-bg-color;
  }
}

.secondary {
  composes: secondary from '../../Button/Button.modules.scss';
  composes: base;

  &:link,
  &:visited {
    color: $button-text-color;
  }
  &:hover {
    color: $secondary-bg-color;
  }
}

.inverted {
  composes: inverted from '../../Button/Button.modules.scss';
  composes: base;

  &:link,
  &:visited {
    color: $color-text;
  }
  &:hover {
    color: $button-text-color;
  }
}
