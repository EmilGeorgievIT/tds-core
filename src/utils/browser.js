/* global document */
/* eslint-disable import/prefer-default-export */
// Provide a way to import the browser's document, which allows it to be mocked in tests

export const doc = document
