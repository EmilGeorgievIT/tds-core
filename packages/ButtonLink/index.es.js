import './dist/index.css'
import ButtonLink from './dist/index.es'

export default ButtonLink
