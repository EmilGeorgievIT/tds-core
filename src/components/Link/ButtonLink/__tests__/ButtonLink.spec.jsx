import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { warn } from '../../../../utils/warn'

import ButtonLink from '../ButtonLink'

jest.mock('../../../../utils/warn', () => (
  {
    warn: jest.fn()
  }
))

describe('ButtonLink', () => {
  const doShallow = (overrides = {}) => shallow(
    <ButtonLink {...overrides}>Go home</ButtonLink>
  )

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders', () => {
    const link = doShallow()

    expect(toJson(link)).toMatchSnapshot()
  })

  it('is an anchor HTML element when using the href attribute', () => {
    const link = doShallow({ href: 'http://telus.com' })

    expect(link).toHaveTagName('a')
    expect(link).toHaveProp('href', 'http://telus.com')
  })

  it('renders a react router link element when passed as a prop', () => {
    const MyLink = () => <span />
    const link = doShallow({ reactRouterLinkComponent: MyLink })

    expect(link).toMatchSelector('MyLink')
  })

  it('must use `reactRouterLinkComponent` and `to` props together', () => {
    const MyLink = () => <span />
    let link = doShallow({ reactRouterLinkComponent: MyLink })

    expect(warn).toHaveBeenCalled()

    jest.clearAllMocks()

    link = doShallow({ to: '/about' })

    expect(link).toHaveProp('to')
    expect(warn).toHaveBeenCalled()
  })

  it('can be presented as one of the allowed variants', () => {
    let button = doShallow()
    expect(button).toHaveClassName('primary')

    button = doShallow({ variant: 'primary' })
    expect(button).toHaveClassName('primary')

    button = doShallow({ variant: 'secondary' })
    expect(button).toHaveClassName('secondary')

    button = doShallow({ variant: 'inverted' })
    expect(button).toHaveClassName('inverted')
  })

  it('passes additional attributes to button element', () => {
    const button = doShallow({ id: 'the-button', tabindex: 1 })

    expect(button).toHaveProp('id', 'the-button')
    expect(button).toHaveProp('tabindex', 1)
  })

  it('does not allow custom CSS', () => {
    const button = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(button).not.toHaveProp('className', 'my-custom-class')
    expect(button).not.toHaveProp('style')
  })
})
