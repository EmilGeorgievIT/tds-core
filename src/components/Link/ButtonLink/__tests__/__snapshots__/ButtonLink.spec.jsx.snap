// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonLink renders 1`] = `
<a
  className="primary"
  href={null}
  to={null}
>
  Go home
</a>
`;
